README.md
# bash-task

A skeleton example of a bash script that simulates a CLI program.


CHANGELOG.md
# Changelog
<!-- markdownlint-disable MD024 -->
<!-- changelog-start -->

## [Unreleased](https://github.com/DonalChilde/bash-task/0.0.1...refs/heads/dev)
<!-- Dont forget to:
    - Update the Unreleased compare version to latest release tag
    - Update compare/_previous_version_tag_
    - Delete <a></a> tag -->
<!-- Copy paste release notes below here -->
<!-- scriv-insert-here -->

## [0.0.1](https://github.com/DonalChilde/bash-task/compare/0.0.0...0.0.1) —  2023-06-15

Whats Changed in 0.0.1

### Added

- Define a template task file.

  - closes
    - [#1](https://github.com/DonalChilde/bash-task/issues/1)

- Add init-venv.sh example.

- Add an example of formatting python code.

  - Pull requests
    - [#5](https://github.com/DonalChilde/bash-task/pull/5)

  - closes
    - [#3](https://github.com/DonalChilde/bash-task/issues/3)

### Changed

- Move the python dev dependencies to a requirements.txt file, as `pip install -e .` is not appropriate for this project.

  - Pull requests
    - [#4](https://github.com/DonalChilde/bash-task/pull/4)

  - closes
    - [#2](https://github.com/DonalChilde/bash-task/issues/2)

## 0.0.0 - 2023-06-11

### Whats Changed in 0.0.0

This is the start of something....

### Added

- Project Start

<https://keepachangelog.com/en/1.0.0/>

<!-- changelog-end -->


src/examples/python-pypi-test.sh
#!/usr/bin/env bash

# This script is based on: https://github.com/adriancooney/Taskfile
# HOME: https://github.com/DonalChilde/bash-task

# Upload a package to PyPi Test

# -e Exit immediately if a pipeline returns a non-zero status.
# -u Treat unset variables and parameters other than the special parameters ‘@’ or ‘*’ as an error when performing parameter expansion.
# -o pipefail If set, the return value of a pipeline is the value of the last (rightmost) command to exit with a non-zero status, or zero if all commands in the pipeline exit successfully.
set -euo pipefail

# Define non-parameter variables here.
# The name of the script
SCRIPT=$0
# The script command line arguments as an array
ARGS=("$@")

# The location of the .pypirc file relative to the project root.
PYPIRC_RELATIVE="secrets/.pypirc"
# The location of the dist directory relative to the project root.
DIST_DIR_RELATIVE="dist/"

function do() {
    # Choose with or without confirmation

    # With confirmation before execution. This will also display the list of commands.
    # Pass in the number of seconds to delay confirmation of commands.
    _do_with_confirmation 5

    # Without confirmation before execution
    # _do_without_confirmation
}

function _define_commands() {
    # Define the commands used in this script.

    # Parameters passed in from the command line are generally assigned
    # to variables in the `_define_variables` function. These variables are then used in
    # the commands. This makes it easier to have a display of the commands in the `--help` function
    # with placeholder names for the CLI parameters.
    COMMANDS=(
        "twine check --strict '$DIST_DIR/*'"
        "twine upload --config-file '$PYPIRC' --repository testpypi '$DIST_DIR/*'"
    )
}

function _define_placeholder_variables() {
    # Define the command variables to be used in the help output.
    PATH_IN="PATH_IN"
    # PATH_OUT="PATH_OUT
    DIST_DIR="PATH_IN/DIST_DIR_RELATIVE"
    PYPIRC="PATH_IN/PYPIRC_RELATIVE"
}

function _define_variables() {
    # Check for correct parameters passed in, and assign to variables to be used in commands.

    # Checks for the presence of an parameter, but not validity. Output help if missing.
    # https://stackoverflow.com/a/25066804
    # Arguments start at 1, as 0 is the `do` or `dry-run` command.
    : ${ARGS[1]?"Missing a required command line argument. run '$SCRIPT --help' for usage instructions."}
    # : ${ARGS[2]?"Missing a required command line argument. run '$SCRIPT help' for usage instructions."}

    # Check for cli arguments in excess of action command - ie. `do` or `dry-run`
    # EXCESS_ARGS=("${ARGS[@]:1}")
    # if [ ${#EXCESS_ARGS[@]} -eq 0 ]; then
    #     # Set default arguments for CODE_PATHS
    #     CODE_PATHS=(./src ./tests)
    # else
    #     CODE_PATHS=(${EXCESS_ARGS[@]})
    # fi

    # Assign variables used in the commands
    # A variable with a default value:
    #   PATH_IN=${ARGS[1]:-"./foo/bar"}
    PATH_IN=$(realpath ${ARGS[1]})
    # PATH_OUT=$(realpath ${ARGS[2]})

    DIST_DIR="$PATH_IN/$DIST_DIR_RELATIVE"
    PYPIRC="$PATH_IN/$PYPIRC_RELATIVE"

}

function _do_without_confirmation() {

    echo "Running Commands"
    echo
    echo
    _run_commands
}

function _do_with_confirmation() {

    echo "This will run the following commands:"
    echo
    dry-run
    echo

    echo "Preparing to upload to PyPi."
    echo "Did you:"
    echo "\t- Check for the correct branch?"
    echo "\t- Update the version number?"
    echo "\t- Update the documentation?"
    echo "\t- Run ALL THE TESTS?"
    echo "\t- Update the changelog?"
    echo "\t- Build a fresh dist/?"
    echo

    # Delay message
    echo "Take $1 seconds to be sure:"
    echo
    _countdown $1
    echo

    # Confirmation dialog
    # https://stackoverflow.com/a/1885534/105844
    read -p "-----Are you sure? (Y/N)-----" -n 1 -r
    echo
    if [[ $REPLY =~ ^[Yy]$ ]]; then
        _run_commands
    else
        echo "Commands Declined"
        exit 1
    fi
}

function _run_commands() {
    _define_variables
    _define_commands

    for i in "${COMMANDS[@]}"; do
        eval $i
    done
}

function dry-run() {
    _define_variables
    _define_commands

    echo "Dry run for $SCRIPT"
    echo
    echo "These are the commands that would be executed."
    echo
    for i in "${COMMANDS[@]}"; do
        echo $i
    done
}

function --help() {
    _define_placeholder_variables
    _define_commands

    HELPTEXT=$(
        cat <<END
    NAME
        $SCRIPT - Upload a package to PyPi.

    SYNOPSIS
        $SCRIPT do PARAMETERS
        $SCRIPT dry-run PARAMETERS
        $SCRIPT --help

    DESCRIPTION
        Uses twine to check and upload a package to PyPI.

    EXAMPLES:
        $SCRIPT do PARAMETERS
            Do the script.

        $SCRIPT dry-run PARAMETERS
            Display the commands that would be run, takes the same paramaters as do.

        $SCRIPT --help
            Output the usage instructions.

    COMMANDS:

END
    )
    printf "$HELPTEXT"
    echo
    for i in "${COMMANDS[@]}"; do
        printf "\n\t$i"
    done
    echo
}

function _countdown() {
    # https://superuser.com/questions/611538/is-there-a-way-to-display-a-countdown-or-stopwatch-timer-in-a-terminal
    # Display a countdown clock.
    # $1 = int seconds
    date1=$(($(date +%s) + $1))
    while [ "$date1" -ge $(date +%s) ]; do
        echo -ne "$(date -u --date @$(($date1 - $(date +%s))) +%H:%M:%S)\r"
        sleep 0.1
    done
}

# Runs the help function if no arguments given to script.
TIMEFORMAT=$'\nTask completed in %3lR'
time "${@:---help}"
################### No code below this line #####################


src/examples/python-clean-test.sh
#!/usr/bin/env bash

# This script is based on: https://github.com/adriancooney/Taskfile
# HOME: https://github.com/DonalChilde/bash-task

# Clean out test and lint artifacts.

# -e Exit immediately if a pipeline returns a non-zero status.
# -u Treat unset variables and parameters other than the special parameters ‘@’ or ‘*’ as an error when performing parameter expansion.
# -o pipefail If set, the return value of a pipeline is the value of the last (rightmost) command to exit with a non-zero status, or zero if all commands in the pipeline exit successfully.
set -euo pipefail

# Define non-parameter variables here.
# The name of the script
SCRIPT=$0
# The script command line arguments as an array
ARGS=("$@")

function do() {
    # Choose with or without confirmation

    # With confirmation before execution. This will also display the list of commands.
    # Pass in the number of seconds to delay confirmation of commands.
    _do_with_confirmation 0

    # Without confirmation before execution
    # _do_without_confirmation
}

function _define_commands() {
    # Define the commands used in this script.

    # Parameters passed in from the command line are generally assigned
    # to variables in the `_define_variables` function. These variables are then used in
    # the commands. This makes it easier to have a display of the commands in the `--help` function
    # with placeholder names for the CLI parameters.
    COMMANDS=(
        "rm -fr $PATH_IN/.tox/"
        "rm -fr $PATH_IN/.nox/"
        "rm -f $PATH_IN/.coverage"
        "rm -fr $PATH_IN/htmlcov/"
        "rm -fr $PATH_IN/.pytest_cache"
        "rm -fr $PATH_IN/.mypy_cache"
    )
}

function _define_placeholder_variables() {
    # Define the command variables to be used in the help output.
    PATH_IN="PATH_IN"
    # PATH_OUT="PATH_OUT
}

function _define_variables() {
    # Check for correct parameters passed in, and assign to variables to be used in commands.

    # Checks for the presence of an parameter, but not validity. Output help if missing.
    # https://stackoverflow.com/a/25066804
    # Arguments start at 1, as 0 is the `do` or `dry-run` command.
    : ${ARGS[1]?"Missing a required command line argument. run '$SCRIPT --help' for usage instructions."}
    # : ${ARGS[2]?"Missing a required command line argument. run '$SCRIPT help' for usage instructions."}

    # Check for cli arguments in excess of action command - ie. `do` or `dry-run`
    # EXCESS_ARGS=("${ARGS[@]:1}")
    # if [ ${#EXCESS_ARGS[@]} -eq 0 ]; then
    #     # Set default arguments for CODE_PATHS
    #     CODE_PATHS=(./src ./tests)
    # else
    #     CODE_PATHS=(${EXCESS_ARGS[@]})
    # fi

    # Assign variables used in the commands
    # A variable with a default value:
    #   PATH_IN=${ARGS[1]:-"./foo/bar"}
    PATH_IN=$(realpath ${ARGS[1]})
    # PATH_OUT=$(realpath ${ARGS[2]})

}

function _do_without_confirmation() {

    echo "Running Commands"
    echo
    echo
    _run_commands
}

function _do_with_confirmation() {

    echo "This will run the following commands:"
    echo
    dry-run
    echo

    # Delay message
    echo "Take $1 seconds to be sure:"
    echo
    _countdown $1
    echo

    # Confirmation dialog
    # https://stackoverflow.com/a/1885534/105844
    read -p "-----Are you sure? (Y/N)-----" -n 1 -r
    echo
    if [[ $REPLY =~ ^[Yy]$ ]]; then
        _run_commands
    else
        echo "Commands Declined"
        exit 1
    fi
}

function _run_commands() {
    _define_variables
    _define_commands

    for i in "${COMMANDS[@]}"; do
        eval $i
    done
}

function dry-run() {
    _define_variables
    _define_commands

    echo "Dry run for $SCRIPT"
    echo
    echo "These are the commands that would be executed."
    echo
    for i in "${COMMANDS[@]}"; do
        echo $i
    done
}

function --help() {
    _define_placeholder_variables
    _define_commands

    HELPTEXT=$(
        cat <<END
    NAME
        $SCRIPT - Clean out test and lint artifacts

    SYNOPSIS
        $SCRIPT do PARAMETERS
        $SCRIPT dry-run PARAMETERS
        $SCRIPT --help

    DESCRIPTION
        Remove tox, coverage, mypy, and pytest artifacts.

    EXAMPLES:
        $SCRIPT do PARAMETERS
            Do the script.

        $SCRIPT dry-run PARAMETERS
            Display the commands that would be run, takes the same paramaters as do.

        $SCRIPT --help
            Output the usage instructions.

    COMMANDS:

END
    )
    printf "$HELPTEXT"
    echo
    for i in "${COMMANDS[@]}"; do
        printf "\n\t$i"
    done
    echo
}

function _countdown() {
    # https://superuser.com/questions/611538/is-there-a-way-to-display-a-countdown-or-stopwatch-timer-in-a-terminal
    # Display a countdown clock.
    # $1 = int seconds
    date1=$(($(date +%s) + $1))
    while [ "$date1" -ge $(date +%s) ]; do
        echo -ne "$(date -u --date @$(($date1 - $(date +%s))) +%H:%M:%S)\r"
        sleep 0.1
    done
}

# Runs the help function if no arguments given to script.
TIMEFORMAT=$'\nTask completed in %3lR'
time "${@:---help}"
################### No code below this line #####################


scripts/init-venv.sh
#!/usr/bin/env bash

# This script is based on: https://github.com/adriancooney/Taskfile
# HOME: https://github.com/DonalChilde/bash-task

# This template mimics a cli program, with --help display, and dry-run display.
# It make it easy to define a list of commands to run, with optional parameters from the command line.
# Optional verification before execution is also available.

# General usage is ./scripts/task.sh do <arguments>
# For a dry-run ./scripts/task.sh dry-run <arguments>
# Usage instructions ./scripts/task.sh --help

# Setup Instructions
#
# 1. Choose with or without confirmation in the `do` function.
# 2. Define the commands in the `_define_commands` function.
# 3. Define the display variables in the `_define_placeholder_variables` function.
# 4. Define the variables used in commands in the `_define_variables`
# 5. Provide usage instructions in the `--help` function

# -e Exit immediately if a pipeline returns a non-zero status.
# -u Treat unset variables and parameters other than the special parameters ‘@’ or ‘*’ as an error when performing parameter expansion.
# -o pipefail If set, the return value of a pipeline is the value of the last (rightmost) command to exit with a non-zero status, or zero if all commands in the pipeline exit successfully.
set -euo pipefail

# Define non-parameter variables here.
# The name of the script
SCRIPT=$0
# The script command line arguments as an array
ARGS=("$@")
# Python version to be used when making virtual environments
PY_VERSION="3.11"

# Export this variable if using pyenv
export PYENV_VERSION="$PY_VERSION"

function do() {
    # Choose with or without confirmation

    # With confirmation before execution. This will also display the list of commands.
    # Pass in the number of seconds to delay confirmation of commands.
    # _do_with_confirmation 5

    # Without confirmation before execution
    _do_without_confirmation
}

function _define_commands() {
    # Define the commands used in this script.

    # Parameters passed in from the command line are generally assigned
    # to variables in the `_define_variables` function. These variables are then used in
    # the commands. This makes it easier to have a display of the commands in the `--help` function
    # with placeholder names for the CLI parameters.
    COMMANDS=(
        "if [ -d '$PROJECT_PATH/.venv' ]; then rm -rf '$PROJECT_PATH/.venv'; fi"
        "python3 -m venv '$PROJECT_PATH/.venv'"
        "source '$PROJECT_PATH/.venv/bin/activate'"
        "export PIP_REQUIRE_VIRTUALENV=true"
        "pip3 install -U pip"
        "pip3 install -r requirements.txt"
    )
}

function _define_placeholder_variables() {
    # Define the command variables to be used in the help output.
    PROJECT_PATH="PROJECT_PATH"
    # PATH_OUT="PATH_OUT
}

function _define_variables() {
    # Check for correct parameters passed in, and assign to variables to be used in commands.

    # Checks for the presence of an parameter, but not validity. Output help if missing.
    # https://stackoverflow.com/a/25066804
    # Arguments start at 1, as 0 is the `do` or `dry-run` command.
    : ${ARGS[1]?"Missing a required command line argument. run '$SCRIPT --help' for usage instructions."}
    # : ${ARGS[2]?"Missing a required command line argument. run '$SCRIPT help' for usage instructions."}

    # Assign variables used in the commands
    # A variable with a default value:
    #   PATH_IN=${ARGS[1]:-"./foo/bar"}
    PROJECT_PATH=$(realpath ${ARGS[1]})
    # PATH_OUT=$(realpath ${ARGS[2]})

}

function _do_without_confirmation() {

    echo "Running Commands"
    echo
    echo
    _run_commands
}

function _do_with_confirmation() {

    echo "This will run the following commands:"
    echo
    dry-run
    echo

    # Delay message
    echo "Take $1 seconds to be sure:"
    echo
    _countdown $1
    echo

    # Confirmation dialog
    # https://stackoverflow.com/a/1885534/105844
    read -p "-----Are you sure? (Y/N)-----" -n 1 -r
    echo
    if [[ $REPLY =~ ^[Yy]$ ]]; then
        _run_commands
    else
        echo "Commands Declined"
        exit 1
    fi
}

function _run_commands() {
    _define_variables
    _define_commands

    for i in "${COMMANDS[@]}"; do
        eval $i
    done
}

function dry-run() {
    _define_variables
    _define_commands

    echo "Dry run for $SCRIPT"
    echo
    echo "These are the commands that would be executed."
    echo
    for i in "${COMMANDS[@]}"; do
        echo $i
    done
}

function --help() {
    _define_placeholder_variables
    _define_commands

    HELPTEXT=$(
        cat <<END
    NAME
        $SCRIPT - Init a Python virtual environment in the project directory.

    SYNOPSIS
        $SCRIPT do PARAMETERS
        $SCRIPT dry-run PARAMETERS
        $SCRIPT --help

    DESCRIPTION
        $SCRIPT is used to init a Python virtual environment, update pip, and install a Python project
            with dependencies as --editable. Expects to be called with the project root dir. Creates a venv
            there at <project_root>/.venv, and also looks in the project root for the pyproject.toml file.
            If a venv already exists, it will be erased and replaced.

    EXAMPLES:
        $SCRIPT do PARAMETERS
            _a_short_description_

        $SCRIPT dry-run PARAMETERS
            Display the commands that would be run, takes the same paramaters as do.

        $SCRIPT --help
            Output the usage instructions.

    COMMANDS:

END
    )
    printf "$HELPTEXT"
    echo
    for i in "${COMMANDS[@]}"; do
        printf "\n\t$i"
    done
    echo
}

function _countdown() {
    # https://superuser.com/questions/611538/is-there-a-way-to-display-a-countdown-or-stopwatch-timer-in-a-terminal
    # Display a countdown clock.
    # $1 = int seconds
    date1=$(($(date +%s) + $1))
    while [ "$date1" -ge $(date +%s) ]; do
        echo -ne "$(date -u --date @$(($date1 - $(date +%s))) +%H:%M:%S)\r"
        sleep 0.1
    done
}

# Runs the help function if no arguments given to script.
TIMEFORMAT=$'\nTask completed in %3lR'
time "${@:---help}"
################### No code below this line #####################


src/task-template.sh
#!/usr/bin/env bash

# This script is based on: https://github.com/adriancooney/Taskfile
# HOME: https://github.com/DonalChilde/bash-task

# This template mimics a cli program, with --help and dry-run display.
# It makes it easy to define a list of commands to run,
# with optional parameters from the command line.
# Confirmation before execution with optional
# delay before execution is also available.

# General usage is ./scripts/task.sh do <arguments>
# For a dry-run ./scripts/task.sh dry-run <arguments>
# Usage instructions ./scripts/task.sh --help

########################################
########## Setup Instructions ##########
########################################

# 1. Set script configuration variables.
# 2. Set custom script variables.
# 3. Edit custom script functions.
# 4. Set --help configuration variables.

##########################################
########## Script configuration ##########
##########################################

# -e Exit immediately if a pipeline returns a non-zero status.
# -u Treat unset variables and parameters other than the special parameters ‘@’ or ‘*’ as an error when performing parameter expansion.
# -o pipefail If set, the return value of a pipeline is the value of the last (rightmost) command to exit with a non-zero status, or zero if all commands in the pipeline exit successfully.
set -euo pipefail

# The script command line arguments as an array
# Do not alter this variable
ARGS=("$@")

# echo commands before eval
ECHO_COMMANDS=1 # true
# ECHO_COMMANDS=0 # false

# confirm execution of commands before eval
EXECUTION_CONFIRMATION=1 # true
# EXECUTION_CONFIRMATION=0 # false

# Seconds to delay execution. Used in conjunction with EXECUTION_CONFIRMATION
DELAY=0

######################################
########## Custom variables ##########
######################################

# None

##########################################
########## --help configuration ##########
##########################################

# The name of the script
SCRIPT=$(basename $0)

# A one line description.
SHORT_DESCRIPTION="A concise one line description of the script"

# A Multiline description of the script.
LONG_DESCRIPTION=$(
    cat <<END
$SCRIPT is used to ... A longer, typically multi line description of the script.
            The second and following lines should be indented 12 spaces to make the --help
            formatting line up.
END
)

# CLI arguments expected, used in the --help output
# CLI_ARGS="PATH_IN PATH_OUT"
CLI_ARGS="PATH_IN"

#############################################
########## custom script functions ##########
#############################################

function _define_commands() {
    # Define the commands used in this script.
    # Variables used in these commands are expected to be defined
    #   - in the script settings above, and/or
    #   - in the `_define_variables` function for actual use, and
    #   - in the `_define_placeholder_variables` function for --help display.

    COMMANDS=(
        "echo 'Command 1'"
        "echo 'Command 2'"
        "ls -la $PATH_IN"
    )
}

function _define_placeholder_variables() {
    # Define the command variables to be used in the --help output.
    # These are variables that are normaly defined from cli arguments,
    # not ones that are already defined in the script settings.

    PATH_IN="PATH_IN"
    # PATH_OUT="PATH_OUT
}

function _define_variables() {
    # Turn arguments from the command line into variables for use in _define_commands
    # This function is called for `do` and `dry-run` but not `--help`.
    # Variables defined here are not visible during --help display.

    # Checks for the presence of an parameter, but not validity. Output help if missing.
    # https://stackoverflow.com/a/25066804
    # Arguments start at 1, as 0 is the `do` or `dry-run` command.
    : ${ARGS[1]?"Missing a required command line argument. run '$SCRIPT --help' for usage instructions."}
    # : ${ARGS[2]?"Missing a required command line argument. run '$SCRIPT help' for usage instructions."}

    # Check for cli arguments in excess of action command - ie. `do` or `dry-run`
    # and set default values if missing.
    # EXCESS_ARGS=("${ARGS[@]:1}")
    # if [ ${#EXCESS_ARGS[@]} -eq 0 ]; then
    #     # Set default arguments for CODE_PATHS
    #     CODE_PATHS=(./src ./tests)
    # else
    #     CODE_PATHS=(${EXCESS_ARGS[@]})
    # fi

    # A variable with a default value:
    #   PATH_IN=${ARGS[1]:-"./foo/bar"}

    PATH_IN=$(realpath ${ARGS[1]})
    # PATH_OUT=$(realpath ${ARGS[2]})
}

function _confirmation_message() {
    # Edit this function for a custom confirmation message.
    echo "This will run the following commands:"
    echo
    _dry-run
    echo
}

##########################################
########## NOT USUALLY MODIFIED ##########
##########################################

function do() {
    if [ $EXECUTION_CONFIRMATION -eq 1 ]; then
        _confirmation_message
        _confirmation_delay
        _confirmation_dialog
    fi
    _run_commands
}

function _confirmation_dialog() {
    # Confirmation dialog
    # https://stackoverflow.com/a/1885534/105844
    read -p "-----Are you sure? (Y/N)-----" -n 1 -r
    echo
    if [[ $REPLY =~ ^[Yy]$ ]]; then
        :
    else
        echo "Commands Declined"
        exit 1
    fi
}

function _confirmation_delay() {
    if [ $DELAY -gt 0 ]; then
        echo "Take $DELAY seconds to be sure:"
        echo
        _countdown $DELAY
        echo
    fi
}

function _run_commands() {
    _define_variables
    _define_commands

    for i in "${COMMANDS[@]}"; do
        if [ $ECHO_COMMANDS -eq 1 ]; then
            echo "$i"
        fi
        eval "$i"
    done
}

function _countdown() {
    # https://superuser.com/questions/611538/is-there-a-way-to-display-a-countdown-or-stopwatch-timer-in-a-terminal
    # Display a countdown clock.
    # $1 = int seconds
    date1=$(($(date +%s) + $1))
    while [ "$date1" -ge $(date +%s) ]; do
        echo -ne "$(date -u --date @$(($date1 - $(date +%s))) +%H:%M:%S)\r"
        sleep 0.1
    done
}

function dry-run() {
    echo "Dry run for $SCRIPT"
    echo
    echo "These are the commands that would be executed."
    echo
    _dry-run
}

function _dry-run() {
    _define_variables
    _define_commands
    for i in "${COMMANDS[@]}"; do
        echo "$i"
    done
}

function --help() {
    _define_placeholder_variables
    _define_commands

    HELPTEXT=$(
        cat <<END
    NAME
        $SCRIPT - $SHORT_DESCRIPTION

    SYNOPSIS
        $SCRIPT do $CLI_ARGS
        $SCRIPT dry-run $CLI_ARGS
        $SCRIPT --help

    DESCRIPTION
        $LONG_DESCRIPTION

    EXAMPLES:
        $SCRIPT do $CLI_ARGS
            Do the script.

        $SCRIPT dry-run $CLI_ARGS
            Display the commands that would be run.

        $SCRIPT --help
            Output the usage instructions.

    COMMANDS:

END
    )
    printf "$HELPTEXT"
    echo
    for i in "${COMMANDS[@]}"; do
        printf "\n\t$i"
    done
    echo
}

function _command_check() {
    if [ "${ARGS[0]}" = "do" ]; then
        :
    elif [ "${ARGS[0]}" = "dry-run" ]; then
        :
    elif [ "${ARGS[0]}" = "--help" ]; then
        :
    else
        echo "The first script argument must be one of ['do', 'dry-run', '--help']."
        echo "Received '${ARGS[0]}'"
        echo "Use '$SCRIPT --help' for usage instructions."
        exit 1
    fi
}

_command_check
# Runs the help function if no arguments given to script.
TIMEFORMAT=$'\nTask completed in %3lR'
time "${@:---help}"
#################################################################
################### No code below this line #####################
#################################################################


scripts/test-bench.sh
#!/usr/bin/env bash

# A place to try out random bash commands

# echo "HI"

# HELLO="Hello"
# HELLO_WORLD="$HELLO world"

# echo "$HELLO world"
# echo "$HELLO_WORLD"
# echo '$HELLO_WORLD'
# echo '"$HELLO_WORLD"'

# command_array=("Hello" "$HELLO world" "$HELLO_WORLD")

# for i in "${command_array[@]}"; do
#     echo $i
#     echo "$i" | grep '\$'
# done
SCRIPT=$0
ARGS=("$@")
function do() {

}

function --help() {
    _default_params
    _define_commands
    echo "Help!"
    printf "\nRunning the following commands:"
    echo
    for i in "${COMMANDS[@]}"; do
        printf "\n\t$i"
    done

}
function _run_commands() {
    _define_params
    _define_commands
    for i in "${COMMANDS[@]}"; do
        eval $i
    done
}
function _define_params() {
    : ${ARGS[1]?"Missing a required command line argument. run '$SCRIPT --help' for usage instructions."}
    PATH_IN=$(realpath ${ARGS[1]})
}
function _default_params() {
    PATH_IN="PATH_IN"
}
function dry_run() {
    _define_params
    _define_commands

    echo "Dry run for $SCRIPT"
    echo
    echo "These are the commands that would be executed."
    for i in "${COMMANDS[@]}"; do
        echo $i
    done

}
function _define_commands() {
    COMMANDS=(
        "echo 'Action_1'"
        "echo 'Action_2'"
        "ls -la $PATH_IN"
    )
}
# Runs the help function if no arguments given to script.
TIMEFORMAT=$'\nTask completed in %3lR'
time "${@:---help}"
################### No code below this line #####################


src/examples/python-pypi-release.sh
#!/usr/bin/env bash

# This script is based on: https://github.com/adriancooney/Taskfile
# HOME: https://github.com/DonalChilde/bash-task

# Upload a package to PyPi

# -e Exit immediately if a pipeline returns a non-zero status.
# -u Treat unset variables and parameters other than the special parameters ‘@’ or ‘*’ as an error when performing parameter expansion.
# -o pipefail If set, the return value of a pipeline is the value of the last (rightmost) command to exit with a non-zero status, or zero if all commands in the pipeline exit successfully.
set -euo pipefail

# Define non-parameter variables here.
# The name of the script
SCRIPT=$0
# The script command line arguments as an array
ARGS=("$@")

# The location of the .pypirc file relative to the project root.
PYPIRC_RELATIVE="secrets/.pypirc"
# The location of the dist directory relative to the project root.
DIST_DIR_RELATIVE="dist/"

function do() {
    # Choose with or without confirmation

    # With confirmation before execution. This will also display the list of commands.
    # Pass in the number of seconds to delay confirmation of commands.
    _do_with_confirmation 5

    # Without confirmation before execution
    # _do_without_confirmation
}

function _define_commands() {
    # Define the commands used in this script.

    # Parameters passed in from the command line are generally assigned
    # to variables in the `_define_variables` function. These variables are then used in
    # the commands. This makes it easier to have a display of the commands in the `--help` function
    # with placeholder names for the CLI parameters.
    COMMANDS=(
        "twine check --strict '$DIST_DIR/*'"
        "twine upload --config-file '$PYPIRC' '$DIST_DIR/*'"
    )
}

function _define_placeholder_variables() {
    # Define the command variables to be used in the help output.
    PATH_IN="PATH_IN"
    # PATH_OUT="PATH_OUT
    DIST_DIR="PATH_IN/DIST_DIR_RELATIVE"
    PYPIRC="PATH_IN/PYPIRC_RELATIVE"
}

function _define_variables() {
    # Check for correct parameters passed in, and assign to variables to be used in commands.

    # Checks for the presence of an parameter, but not validity. Output help if missing.
    # https://stackoverflow.com/a/25066804
    # Arguments start at 1, as 0 is the `do` or `dry-run` command.
    : ${ARGS[1]?"Missing a required command line argument. run '$SCRIPT --help' for usage instructions."}
    # : ${ARGS[2]?"Missing a required command line argument. run '$SCRIPT help' for usage instructions."}

    # Check for cli arguments in excess of action command - ie. `do` or `dry-run`
    # EXCESS_ARGS=("${ARGS[@]:1}")
    # if [ ${#EXCESS_ARGS[@]} -eq 0 ]; then
    #     # Set default arguments for CODE_PATHS
    #     CODE_PATHS=(./src ./tests)
    # else
    #     CODE_PATHS=(${EXCESS_ARGS[@]})
    # fi

    # Assign variables used in the commands
    # A variable with a default value:
    #   PATH_IN=${ARGS[1]:-"./foo/bar"}
    PATH_IN=$(realpath ${ARGS[1]})
    # PATH_OUT=$(realpath ${ARGS[2]})

    DIST_DIR="$PATH_IN/$DIST_DIR_RELATIVE"
    PYPIRC="$PATH_IN/$PYPIRC_RELATIVE"

}

function _do_without_confirmation() {

    echo "Running Commands"
    echo
    echo
    _run_commands
}

function _do_with_confirmation() {

    echo "This will run the following commands:"
    echo
    dry-run
    echo

    echo "Preparing to upload to PyPi."
    echo "Did you:"
    echo "\t- Check for the correct branch?"
    echo "\t- Update the version number?"
    echo "\t- Update the documentation?"
    echo "\t- Run ALL THE TESTS?"
    echo "\t- Update the changelog?"
    echo "\t- Build a fresh dist/?"
    echo

    # Delay message
    echo "Take $1 seconds to be sure:"
    echo
    _countdown $1
    echo

    # Confirmation dialog
    # https://stackoverflow.com/a/1885534/105844
    read -p "-----Are you sure? (Y/N)-----" -n 1 -r
    echo
    if [[ $REPLY =~ ^[Yy]$ ]]; then
        _run_commands
    else
        echo "Commands Declined"
        exit 1
    fi
}

function _run_commands() {
    _define_variables
    _define_commands

    for i in "${COMMANDS[@]}"; do
        eval $i
    done
}

function dry-run() {
    _define_variables
    _define_commands

    echo "Dry run for $SCRIPT"
    echo
    echo "These are the commands that would be executed."
    echo
    for i in "${COMMANDS[@]}"; do
        echo $i
    done
}

function --help() {
    _define_placeholder_variables
    _define_commands

    HELPTEXT=$(
        cat <<END
    NAME
        $SCRIPT - Upload a package to PyPi.

    SYNOPSIS
        $SCRIPT do PARAMETERS
        $SCRIPT dry-run PARAMETERS
        $SCRIPT --help

    DESCRIPTION
        Uses twine to check and upload a package to PyPI.

    EXAMPLES:
        $SCRIPT do PARAMETERS
            Do the script.

        $SCRIPT dry-run PARAMETERS
            Display the commands that would be run, takes the same paramaters as do.

        $SCRIPT --help
            Output the usage instructions.

    COMMANDS:

END
    )
    printf "$HELPTEXT"
    echo
    for i in "${COMMANDS[@]}"; do
        printf "\n\t$i"
    done
    echo
}

function _countdown() {
    # https://superuser.com/questions/611538/is-there-a-way-to-display-a-countdown-or-stopwatch-timer-in-a-terminal
    # Display a countdown clock.
    # $1 = int seconds
    date1=$(($(date +%s) + $1))
    while [ "$date1" -ge $(date +%s) ]; do
        echo -ne "$(date -u --date @$(($date1 - $(date +%s))) +%H:%M:%S)\r"
        sleep 0.1
    done
}

# Runs the help function if no arguments given to script.
TIMEFORMAT=$'\nTask completed in %3lR'
time "${@:---help}"
################### No code below this line #####################
